### DISCUSSION TITLE
Instructions for a static build for Linux on Ubuntu 22.04, and then compiling a sample C++ file

### DISCUSSION THREAD

**Discussion by dreddi** [MTG/essentia#1295] [General]
Current as of 4 Jan 2023, commit https://github.com/MTG/essentia/commit/02d8a6fab0a5c4cdc2990624327efd96899cd237

Refer to these for the original instructions:

- https://essentia.upf.edu/FAQ.html#linux-macos-static-builds
- https://essentia.upf.edu/FAQ.html#how-to-compile-my-own-c-code-that-uses-essentia 

And then refer below for troubleshooting guidance.

## Issue 1: the qt4 dependency will not build on gcc-11, but it is known to work on gcc-8
Ref: https://bugreports.qt.io/browse/QTBUG-81361
Solution: Compile and use gcc-8, as it is not in the Ubuntu 22 repositories.
```
mkdir gcc-8
cd gcc-8
# download and extract https://mirror.fcix.net/gnu/gcc/gcc-8.5.0/gcc-8.5.0.tar.xz 
# to ./gcc-8.5.0/

cd ./gcc-8.5.0/
contrib/download_prerequisites
mkdir ../tmp/
mkdir ../build/
cd ../tmp/
../gcc-8.5.0/configure --prefix=/full/path/to/our/working/directory/gcc-8/build/
make bootstrap
make install #takes hours, could be sped up with some threading option somewhere in make

#
# We're going to add gcc-8 to our compiler switcher
#

# First adding the ones already on my machine (assuming they're not already on this list)
# see sudo update-alternatives --list g++; sudo update-alternatives --list gcc
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-11 20
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 20 
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 15
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 15 

# adding our new gcc-8 compilers to this list
sudo update-alternatives --install /usr/bin/g++ g++ /full/path/to/our/working/directory/gcc-8/build/bin/g++ 10
sudo update-alternatives --install /usr/bin/gcc gcc /full/path/to/our/working/directory/gcc-8/build/bin/gcc 10 


# choose our gcc-8 compiler to become our system wide default (you can change this back after with the same command)
sudo update-alternatives --config g++; sudo update-alternatives --config gcc
```

## Issue 2: While building the dependencies, ./waf may throw errors when run in python3, so we must force python2.
```
sudo apt install python2

# untested: replaces all mentions of ./waf with python2 ./waf in the build_xxx.sh scripts
cd ~/essentia/packaging/debian_3rdparty/
sed "s,^./waf,python2 ./waf," build_*
```

Now build the dependencies, including gaia (the sample standard_mfcc.cpp requires it)

```
cd ../../
packaging/build_3rdparty_static_debian.sh --with-gaia 

# build chromaprint which is also required by the sample cpp file
packaging/debian_3rdparty/build_chromaprint.sh
```

Now build Essentia.

## Issue 3: Essentia will not build in gcc-8.
Solution: gcc-11 works!

```
# Choose gcc-11 if you have it installed
sudo update-alternatives --config g++; sudo update-alternatives --config gcc


python2 ./waf configure  --with-static-examples
python2 ./waf
```

## Issues 4 & 5: The provided sample command must be modified to suit a static build (update the absolute paths to suit your system). Finally, we must add -lchromaprint to the command for it to build, and link order matters! 

```
# From wherever you choose to save the sample cpp file standard_mfcc.cpp, run 
g++ -pipe -Wall -O2 -fPIC \
-I/home/user/essentia/lib/essentia/src \
-I/home/user/essentia/lib/essentia/src/essentia/scheduler \
-I/home/user/essentia/lib/essentia/src/essentia/streaming  \
-I/home/user/essentia/lib/essentia/src/essentia/utils \
-I/home/user/essentia/lib/essentia/packaging/debian_3rdparty/include/taglib \
-I/home/user/essentia/lib/essentia/packaging/debian_3rdparty/include/gaia2 \
-I/home/user/essentia/lib/essentia/packaging/debian_3rdparty/include \
-I/home/user/essentia/lib/essentia/packaging/debian_3rdparty/include/QtCore \
-D__STDC_CONSTANT_MACROS standard_mfcc.cpp -o standard_mfcc \
-L/home/user/essentia/lib/essentia/packaging/debian_3rdparty/lib \
-L/home/user/essentia/lib/essentia/build/src \
 -lessentia -lchromaprint -lfftw3 -lyaml -lavresample -lavformat -lavcodec -lavutil -lsamplerate -ltag -lfftw3f -lQtCore -lgaia2 
```

Done, it compiles with no further output.

**Comment by regorxxx** [comment]
Could you please, try making a build for windows (with Gaia)?
https://github.com/MTG/essentia/discussions/1279

Some people from foobar2000 and me have been trying to replace the AcousticBrainz binaries with new ones including GAIA, so people can continue using the models and build their own database, and it's essentially impossible. 

I have already created an issue, asked to AcousticBrainz people (they will not help), wrote to the developers, ... no one cares about it, even when there are dozen of users creating issues from time to time pointing to build problems and the current guide not working.
